### PR TITLE
Wrap the various crt0 objects into static libraries.

### PR DESCRIPTION
**Pull request by statham-arm**
When linking an application against picolibc, if I want to choose between its various choices of `crt0-foo.o` objects (say, `crt0.o` vs `crt0-semihost.o`), the only way I know to do it is to put the appropriate object file name on the linker command line. This requires knowing the _absolute path_ of that file, which is unergonomic: it involves automated build systems having to find out that path and write it into makefiles, and worse, manually written makefiles might have to hard-code the pathname and then not be able to run on a machine where the toolchain is installed in a different directory.

It would be nicer to be able to write a short option that just says 'use `crt0-semihost.o`, please' and have the linker automatically look in `$sysroot/lib` to find it.

The best answer I've found is to have each crt0 object be wrapped in a tiny library file, so that instead of including a full path to `crt0-foo.o` on the command line, I can just write `-lcrt0foo`. I don't know if this works in all linkers, but with lld at least it seems fine: lld already knows it wants to find a symbol called _start, and if that symbol isn't among its explicit object files, it will search the libraries for it and find it in `libcrt0foo.a`.

This commit does not remove, or prevent installation of, the bare `crt0.o` and friends. The new libraries are an alternative, but the usual approach should still work too.

### CODE REVIEW

**Comment by jnohlgard** [comment]
There is a custom option in `picolibc.specs` for `-crt0=xyz` to pick a custom crt0, e.g. `-crt0=semihost`. It uses the special tag `%s` to tell gcc to look for the crt0 object in the default directories for start files. See https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gcc/Spec-Files.html for more detailed docs.

Does this option help in you use case?

**Comment by rdiez** [comment]
I am a little nervous about such crt0 objects. When writing embedded software, the best crt0 is actually empty. This is what I am using now:

https://github.com/rdiez/DebugDue/blob/master/Project/src/BareMetalSupport/crt0.cpp

I cannot see "choices of crt0-foo.o objects (say, crt0.o vs crt0-semihost.o)" at the moment. There seems to be one crt0 per architecture, like picocrt/machine/arm/crt0.c , but with conditionals inside like "#ifdef CRT0_SEMIHOST".

picolibc started by dropping libgloss, and this feels like a covert attempt to bring ligbloss back. Those crt0 files make assumptions about the runtime environment which may not hold true. Even in the semihosting scenario, your runtime environment may depend on the simulator you are using, and how you have configured it.

I would like to build picolibc for one architecture, and use if for both the real hardware and the simulator. A crt0 file with "#ifdef CRT0_SEMIHOST" seems to contradict that idea.

I think picolibc should build a generic library. If you then need start-up code and linker script files for different target configurations, then maybe a separate, explicit libgloss library would be a better choice.

**Comment by statham-arm** [comment]
@jnohlgard , thanks for the suggestion about `picolibc.specs`. But no, I don't think it can help me, because as far as I can tell, it's only understood by the genuine GNU tools. I'm using picolibc with a toolchain composed of `clang` and `lld`, which don't have any support for specs files at all that I can see (other than occasionally forwarding the option to gcc, according to one unit test).

@rdiez:

> I cannot see "choices of crt0-foo.o objects (say, crt0.o vs crt0-semihost.o)" at the moment. There seems to be one crt0 per architecture, like picocrt/machine/arm/crt0.c , but with conditionals inside like "#ifdef CRT0_SEMIHOST".

Yes, but the `build.meson` file I'm modifying will build more than one _object_ file, by compiling the same source with different choices of those `#ifdef`s you mention. The result is that the output lib directory contains multiple `crt0.o`, `crt0-semihost.o`, etc. Once you have that in your install directory, you need a way to choose which one to link a given program with.

> this feels like a covert attempt to bring ligbloss back

I don't really know anything about libgloss – I didn't come to picolibc from previous use of newlib. So if this PR is a covert attempt to bring it back, it's so covert that even I don't know about it!

**Comment by rdiez** [comment]
I do not think it is a good idea to build several crt0-xxx.o object files out of one crt0.c file.

The file I am using (which I linked above) states that GCC (or the GNU linker) hard-cores the name crt0.o , so using other names will probably not work with a GCC-based toolchain. I am using an empty crt0.o in order to overcome this limitation. This way, you can implement the start-up code in a file with another name.

Would it be possible to split the crt0.c files, so that they do not need the #if statements inside? That probably wouldn't be enough, but it maybe cleaner anyway.

I hope one of the build options will be to not build any crt0.o at all. That shouldn't be a default part of a C runtime library, only something that the user application chooses to use depending on the target. I even think that having a default crt0.o in the search path is risky. I would rather make explicit which crt0 from picolibc I want to use, if any of the supplied crt0 files matches my start-up needs. I normally have my own start-up code customized for each target, but when running self-tests under a simulator, maybe a generic crt0 provided by picolibc would save some work.

**Comment by rdiez** [comment]
> > I do not think it is a good idea to build several crt0-xxx.o object files out of one crt0.c file.
> 
> It's just a source code management issue; separate .c files could share common implementation bits from a shared header file.
> [...]
> They're all closely related though, and I'd prefer to avoid duplicating source in a bunch of places.

Let's look specifically at file picocrt/machine/arm/crt0.c . The bottom half of it is under "#ifdef CRT0_SEMIHOST", and that section starts like this, with a bunch of #include's:

#ifdef CRT0_SEMIHOST

#include <semihost.h>
#include <unistd.h>
#include <stdio.h>

That feels like it should live in a separate source file.


> You can use the gcc option `-nostartfiles` to elide crt0.o from the link

If that's so easy, why keep the name crt0?

After decades of bad GCC/binutils habits, the name crt0.o is deeply engrained in the muscle memories of countless developers. Many will assume that crt0.o is always linked into the application, because the linker has that name hard-coded inside. Most people will probably avoid thinking about unusual compiler flags, or the black art of GCC spec files, which may change this rule.

So, if you are really thinking about names like "crt0.o vs crt0-semihost.o", many people will think first that crt0.o is always there, and crt0-semihost.o is an optional extra.

There is no need to support such confusion. Just name the files picocrt-init-normal.c (if there is something 'normal' at all in the embedded world), picocrt-init-semihost.c and the like.


The way you describe '-nostartfiles' is not accurate anyway. This option is not about crt0.o, but about many other things. When I started with GCC toolchains for embedded ARM, I had a hard time and tried everything. I think '-nostartfiles' did not work for me, that is why I ended up adding an empty crt0 to my firmware project. GCC's documentation for '-nostartfiles' is rather vague, which is a recipe for chaos across all supported architectures. The only option specific for crt0.o is '-mno-crt0', but that option is unfortunately not supported by many targets.

File: picocrt/meson.build [hunk 1]
```diff
@@ -102,6 +109,13 @@ foreach target : targets
 	       _crt.extract_objects(src_picocrt)
 	      )
 
+  static_library('crt0hosted',
```
Comment (keith-packard) [inline picocrt/meson.build:112]: You'll need to deal with multilib just like the crt0*.o files. That's probably most easily done by removing the .o from the `crt_name`, `crt_hosted_name` etc values and appending the `.o` back on for the executable targets and using it plain for the static library targets.

**Review by keith-packard — CHANGES_REQUESTED** [review]
See the comment for line 104 in picocrt/meson.build about supporting multilib.

File: picocrt/meson.build (hunk 1 above)
Comment (statham-arm) [reply picocrt/meson.build]: Thanks for the suggestion. Also this ensures the library and object file names are exactly consistent with each other, which in my previous approach they accidentally weren't!

**Comment by rdiez** [comment]
I would at least rename 'crt0' to 'crt0-freestanding' to prevent the confusion I talked about above.

> Yes, but the meson.build file I'm modifying will build more than one object file,
> by compiling the same source with different choices of those #ifdefs you mention.

Does that mean that the current build script only builds one variant, and the modified one will always build all 4?

Does that imply that all architectures will have exactly 4 variants?

Wouldn't it be better to pass an argument with a list of all variants you want to build?

I also noticed other things when looking around:

1) I wouldn't have different source files with the same name it they may used at the same time in one build:

picocrt/shared/crt0.c
picocrt/machine/arm/crt0.c

2) I saw the following comment in picocrt/meson.build source code:

# The 'hosted' variant calls 'exit' after return from main (c lingo: hosted execution environment)

Where is that happening? I am looking at picocrt/machine/arm/crt0.c , and the only code paths which call _exit are under "#ifdef CRT0_SEMIHOST".

**Comment by statham-arm** [comment]
@rdiez:
> Does that mean that the current build script only builds one variant, and the modified one will always build all 4?

No. The current build script, before this PR, already _unconditionally_ builds three crt0 object files: `crt0.o`, `crt0-hosted.o` and `crt0-minimal.o`. Additionally, a fourth object `crt0-semihost.o` is built if `has_arm_semihost` is defined.

My change in this PR causes a library to be built alongside each of those object files, if and only if that object file is built. I'm not making any changes to which object files are built in the first place.

Your comments in this thread mostly seem to be critiquing the existing code that's the context for my change, and not the change itself. If you'd like the whole crt0 system redesigned, perhaps a separate PR or issue might be a good place to discuss that, to keep it separate from this minor ergonomics improvement to the current scheme?

**Review by keith-packard — APPROVED** [review]
This looks like it should work fine. We don't (currently) have any tests for this change, so if you want to consider how to test it using the existing CI system, that could be a nice follow-on PR.

**Comment by keith-packard** [comment]
ok, I gave this a test and discovered that only the default multilib target ended up with 'lib' included in the filename. If you'll check out newlib/meson.build, you'll see that it explicitly adds 'lib' to each of the path names other than the default. That's due to a quirk in meson which I still haven't managed to get them to fix upstream -- picolibc is including directory names in the target names, which meson doesn't officially support, but which is required to get them installed with the correct names.

I think that means you need to create separate names for the .o and .a versions so that you can insert 'lib' in the right place. Something like
```
    crt_name = join_paths(target, 'crt0').o
    libcrt_name = join_paths(target, 'libcrt0')
```

You will probably need to test this using gcc so that you get multilib configurations.

**Comment by statham-arm** [comment]
Yes, I was able to reproduce the problem you described using `do-x86-configure`. I think this version fixes it.

Incidentally, I had unexpected behaviour at install time during that test. I used `-Dprefix` to try to make the library install in a temporary directory of my user filespace, and _most_ files were installed there, but `ninja install` for some reason still tried to install `picolibc.specs` into `/usr/lib`, via meson's automatic attempt to elevate privilege via pkexec. I'd usually expect when manually specifying an install prefix that _everything_ is installed into that prefix. I found the extra `specsdir` setting I could use to turn that off, but is that worth a warning in `build.md`, perhaps?

**Comment by keith-packard** [comment]
> Ah, now I think I see what's going on. It's because `meson.build` starts by running `cc -print-search-dirs` to find the install directory `cc_install_dir`, and treats that as a higher-priority option for `specs_dir` than the user-provided `prefix`.

Yeah, I realized I didn't think about my previous response -- it's not really a bug in the code as the intent is to install `picoilbc.specs` where the gcc will search for it so that you can use `gcc -specs=picolibc.specs` and have the compiler completely configured for picolibc use. I was thinking you were using `DESTDIR=`, which should be respected for all of the installed files.

> For this test purpose I built picolibc using Ubuntu's ordinary system gcc, which is probably not useful for any purpose more profound than checking the names of output files! Perhaps this priority order would have made more sense in actually sensible use cases. :-)

Yup. When you install picolibc for use with an embedded gcc configuration, like arm-none-eabi-gcc, then having picolibc.specs get installed right into the compiler itself it convenient. I'll consider this a gap in documentation though; it should be clear where everything gets installed, and it's not clear that picolibc.specs gets installed in a pretty unexpected location.

**Comment by rdiez** [comment]
No mention of the other variants crt0-hosted, crt0-minimal and crt0-semihost, which are also built?

**Comment by keith-packard** [comment]
Thanks for the doc update.

**Pull request merged**